src/mpyl/utilities/docker/__init__.py

```diff
@@ -56,7 +56,20 @@ def from_dict(config: Dict):
             raise KeyError(f'Docker config could not be loaded from {config}') from exc
 
 
-def stream_docker_logging(logger: Logger, generator, task_name: str, level=logging.INFO) -> None:
+def execute_with_stream(container: Container, command: list[str]) -> Iterable[tuple[str, bytes]]:
+    return cast(Iterable[tuple[str, bytes]], container.execute(command=command, stream=True))
+
+
+def stream_encoded_logging(logger: Logger, generator: Iterable[tuple[str, bytes]], task_name: str,
+                           level=logging.INFO) -> None:
+    def decode():
+        for _origin, value in generator:
+            yield value.decode(errors="replace")
+
+    stream_docker_logging(logger, decode(), task_name, level)
+
+
+def stream_docker_logging(logger: Logger, generator: Iterator[str], task_name: str, level=logging.INFO) -> None:
```

Comment:
You could extend the generator type `Iterator[str]` to include `Iterable[tuple[str, bytes]]` and add map the type down to `Iterator[str]` if necessary?

Reply:
You mean you want to combine it with the function above? I thought it was pretty clean like this, by forcing people to use the correct function based on the type of generator they have.

Reply:
Yeah, but unfortunately it's not obvious from the 🐳 API which of the generators people are dealing with. What about something like:
```python
def stream_docker_logging(logger: Logger, generator: Union[Iterator[str], Iterator[tuple[str, bytes]]], task_name: str,
                          level=logging.INFO) -> None:
    while True:
        try:
            next_item = next(generator)
            log_line = next_item[1].decode(errors="replace") if isinstance(next_item, tuple) else next_item
            logger.log(level, Text.from_ansi(log_line))
        except StopIteration:
            logger.info(f'{task_name} complete.')
            break
```